### ISSUE TITLE
Request to support .HKL file format as input option

### ISSUE THREAD

**Issue by gyuhyeokcho** [rs-station/careless#94]
It would be greatly appreciated if Careless takes .HKL file format as an input option.

As an XDS user for data processing, to use Careless, I have to convert output XDS_ASCII.HKL or temp.ahkl files to .mtz using CCP4 Pointless, selecting the "Just combine input files" option. While this approach works for most XDS CORRECT outputs (XDS_ASCII.HKL), it fails for merged datasets like XSCALE outputs (temp.ahkl). The failure occurs because Pointless improperly processes the ITEM_ISET column, resulting in merging of multiple reflections with identical HKL values originating from different datasets into a single HKL reflection. To address this issue, I manually adjusted the frame numbers by adding arbitrary values to original datasets (e.g., 1 to 1001, 360 to 1360). However, this workaround is not a robust solution.

To eliminate this issue, could Careless be updated to accept .HKL files as input? This enhancement would eliminate the need for the Pointless conversion step and allow for a more efficient data merging process, and also benefit XDS users.

Thanks,

Gyuhyeok

**Comment by kmdalton** [comment]
Hi @gyuhyeokcho,

I will consider adding HKL support to the careless CLI. However, the HKL format is not standardized and even the INTEGRATE.HKL and XDS_ASCII.HKL files produced by XDS have different header formats. I don't really know how to support it in a maintainable way. I would like a little more info about your use case before I decide how to deal with it. I see you mentioned XSCALE outputs and XDS_ASCII.HKL filles. Can I ask why you are providing these already scaled inputs to careless? Would it not make more sense to use the unscaled reflections from INTEGRATE.HKL? 

Would it be okay to provide an external conversion script with syntax like?  
```
careless.hkl2mtz INTEGRATE.HKL unmerged.mtz
```
I could implement something like this relatively easily. Let me know your thoughts.

**Comment by gyuhyeokcho** [comment]
Dear @kmdalton,

The external conversion script would be perfect!

In my usual data processing workflow, I scale the data in XDS using either CORRECT or XSCALE. Subsequently, I apply anisotropic cutoff using STARANISO, which re-determines the image scales using an already scaled file. I thought a similar approach could be applied to Careless as well.

I agree that it makes more sense to use unscaled data for Careless to newly determine scale factors. Although I have tested only a single dataset, I compared the results obtained from Careless using either CORRECT or INTEGRATE outputs. Interestingly, the CORRECT outputs yielded better results. So, I decided to continue using CORRECTed files as input for Careless.

**Comment by kmdalton** [comment]
Okay --  sounds like I should support both `INTEGRATE.HKL` and `XDS_ASCII.HKL`. I'll try to get to this later this week. Thanks for sharing your experience processing XDS data.

**Comment by gyuhyeokcho** [comment]
Thanks alot!

**Comment by kmdalton** [comment]
Hi @gyuhyeokcho,

Can you share an example XSCALE output? I don't have one handy. What is the default file name? Can you paste the header here?

Thanks,
Kevin

**Comment by kmdalton** [comment]
I did an implementation of `careless.xds2mtz`. You may test it out by installing the `xds` branch:

```bash
conda create -yn careless-xds python=3.10
conda activate careless-xds
pip install --upgrade pip
pip install git+https://github.com/rs-station/careless@xds
```

I will merge it into main this week and do a new release. Let me know if you need any alterations for your use case.

**Comment by gyuhyeokcho** [comment]
Hi @kmdalton,

Here is one of the XSCALE header. The default file name is "temp.ahkl".

```
!FORMAT=XDS_ASCII    MERGE=FALSE    FRIEDEL'S_LAW=FALSE
!DATE= 3-Feb-2023
!Generated by XSCALE   (VERSION Jan 10, 2022  BUILT=20220220)
!OUTPUT_FILE=temp.ahkl                                        
!COMPRISES THE FOLLOWING SCALED INPUT FILES:
! ISET=      1 INPUT_FILE=/home/Structures/data1/XDS_ASCII.HKL
! ISET=      2 INPUT_FILE=/home/Structures/data2/XDS_ASCII.HKL
! ISET=      1 X-RAY_WAVELENGTH=   0.97957 (<0 if unknown)
! ISET=      2 X-RAY_WAVELENGTH=   0.97957 (<0 if unknown)
! ISET=      1 UNIT_CELL_CONSTANTS=    139.18    139.18    176.36  90.000  90.000  90.000
! ISET=      2 UNIT_CELL_CONSTANTS=    139.30    139.30    176.13  90.000  90.000  90.000
! ISET=      1 MERGE=FALSE FRIEDEL'S_LAW=FALSE
! ISET=      2 MERGE=FALSE FRIEDEL'S_LAW=FALSE
!
!SPACE_GROUP_NUMBER=   94
!UNIT_CELL_CONSTANTS=    139.18    139.18    176.36  90.000  90.000  90.000
!NUMBER_OF_ITEMS_IN_EACH_DATA_RECORD=  10
!ITEM_H=1
!ITEM_K=2
!ITEM_L=3
!ITEM_IOBS=4
!ITEM_SIGMA(IOBS)=5
!ITEM_XD=6
!ITEM_YD=7
!ITEM_ZD=8
!ITEM_PSI=9
!ITEM_ISET=10
!END_OF_HEADER
     0     0     4  1.151E+04  2.882E+04  1518.7  1586.2  1175.5  -41.11 2
     0     0    -6  5.355E+08  2.241E+07  1613.3  1544.9   355.0 -138.40 1
     0     0    -6  5.386E+08  2.254E+07  1614.5  1791.1   177.2   40.48 1
     0     0     6  5.569E+08  2.331E+07  1500.5  1791.5   357.1  139.28 1
     0     0     6  5.663E+08  2.370E+07  1499.5  1545.3   175.1  -41.36 1
     0     0    -6  5.427E+08  2.829E+07  1614.6  1791.3  1177.2   40.37 2
     0     0     6  5.527E+08  2.880E+07  1499.5  1545.2  1175.1  -41.26 2
     0     0    -8  1.056E+07  5.843E+05  1633.6  1832.1   177.6   40.33 1
     0     0    -8  1.009E+07  5.661E+05  1632.2  1503.8   354.6 -138.25 1
     0     0     8  1.040E+07  5.786E+05  1480.3  1504.3   174.8  -41.51 1
     0     0     8  1.050E+07  5.792E+05  1481.7  1832.7   357.4  139.43 1
     0     0    -8  1.171E+07  6.669E+05  1633.8  1832.3  1177.6   40.23 2
     0     0     8  1.258E+07  7.161E+05  1480.3  1504.1  1174.8  -41.40 2
     0     0   -10  1.962E+07  1.005E+06  1651.0  1462.6   354.3 -138.11 1
```

Thanks,

Gyuhyeok

**Comment by kmdalton** [comment]
Thanks, @gyuhyeokcho. I added your `ahkl` file to my tests. I'll merge this later today. Expect a new release later this week. Please feel free to reopen this issue if you find something not working.

**Comment by DorisMai** [comment]
I tried using careless.xds2mtz by
`careless.xds2mtz INTEGRATE.HKL integrate_careless.mtz`

and then running careless by
```
 careless mono \
--iterations=10 \
--merge-half-datasets \
--test-fraction=0.05 \
--disable-image-scales \
--mlp-layers=20 \
"BATCH,dHKL,XDET,YDET" \
integrate_careless.mtz \
myoutput_careless
```

I encountered the following error:
```
Careless version 0.3.0
Traceback (most recent call last):
  File "/opt/anaconda3/envs/careless/bin/careless", line 8, in <module>
    sys.exit(main())
  File "/opt/anaconda3/envs/careless/lib/python3.8/site-packages/careless/careless.py", line 9, in main
    run_careless(parser)
  File "/opt/anaconda3/envs/careless/lib/python3.8/site-packages/careless/careless.py", line 30, in run_careless
    inputs,rac = df.format_files(parser.reflection_files)
  File "/opt/anaconda3/envs/careless/lib/python3.8/site-packages/careless/io/formatter.py", line 138, in format_files
    return self((load(f) for f in files))
  File "/opt/anaconda3/envs/careless/lib/python3.8/site-packages/careless/io/formatter.py", line 115, in __call__
    data, rac = self.get_data_and_asu_collection(datasets)
  File "/opt/anaconda3/envs/careless/lib/python3.8/site-packages/careless/io/formatter.py", line 78, in get_data_and_asu_collection
    ds = self.prep_dataset(ds, sg)
  File "/opt/anaconda3/envs/careless/lib/python3.8/site-packages/careless/io/formatter.py", line 298, in prep_dataset
    raise ValueError(
ValueError: No matching uncertainty key found for intensity key IOBSplease manually specify the uncertainty key. 
```

**Comment by kmdalton** [comment]
It wants you to specify the key for the error estimates for IOBS. I think the flag is `--uncertainty-key` Can you share the output of `rs.mtzdump integrate_careless.mtz`?

**Comment by DorisMai** [comment]
```
Spacegroup: P212121
Extended Hermann-Mauguin name: P 21 21 21
Unit cell dimensions: 5.246 10.417 21.016 90.000 90.000 90.000

mtz.head():

          IOBS  SIGMA  XCAL   YCAL  ZCAL     RLP     PEAK  CORR  ...  ALF0  BET0   ALF1  BET1   PSI  ISEG  BATCH  PARTIAL
H  K  L                                                          ...                                                     
-6 -2 -3 76.39  37.17 207.9 1624.5  61.0  0.0243 98.49495    14  ... 79.47  0.06 142.37  1.71 16.35   1.0     61    False
      -2 952.4  50.36 192.5 1595.1  62.8 0.02472    100.0    65  ... 79.47  0.06 144.21   1.7  8.99   1.0     63    False
      -1 52.16  37.11 176.5 1565.6  64.5 0.02516 99.99992     7  ... 79.47  0.06 146.09  1.69  2.36   1.0     64    False
      0  138.7  37.04 160.2 1536.6  66.1 0.02562 96.94247    23  ... 79.32  0.06 147.97  1.69 -3.46   1.0     66    False
      1  101.5  38.81 143.2 1507.5  67.7 0.02609    100.0    17  ... 79.24  0.06 149.88  1.69 -8.58   1.0     68    False

[5 rows x 20 columns]

mtz.describe():

            IOBS     SIGMA      XCAL      YCAL      ZCAL  ...      ALF1      BET1       PSI    ISEG     BATCH
count   5076.000  5076.000  5076.000  5076.000  5076.000  ...  5076.000  5076.000  5076.000  5076.0  5076.000
mean    1045.649    44.909  1024.228  1036.429    66.132  ...     5.703     1.266    -0.257     1.0    64.821
std     1867.077    33.771   640.855   448.673    37.975  ...   108.529     0.329    77.021     0.0    38.097
min      -73.120     1.392    13.500    35.400     0.200  ...  -179.910     0.040  -179.470     1.0     0.000
25%      129.900    30.245   408.700   687.875    33.175  ...   -65.775     1.070   -52.380     1.0    32.000
50%      385.750    37.370  1023.100  1037.650    66.200  ...     6.210     1.340    -0.590     1.0    65.000
75%     1126.250    46.062  1640.450  1384.175    99.125  ...   110.715     1.530    52.335     1.0    98.000
max    21880.000   444.500  2034.400  2015.500   131.900  ...   179.980     1.750   179.970     1.0   132.000

[8 rows x 19 columns]

mtz.dtypes:

IOBS       Intensity
SIGMA         Stddev
XCAL         MTZReal
YCAL         MTZReal
ZCAL         MTZReal
RLP          MTZReal
PEAK         MTZReal
CORR          MTZInt
MAXC          MTZInt
XOBS         MTZReal
YOBS         MTZReal
ZOBS         MTZReal
ALF0         MTZReal
BET0         MTZReal
ALF1         MTZReal
BET1         MTZReal
PSI          MTZReal
ISEG       Intensity
BATCH          Batch
PARTIAL         bool
dtype: object
```

**Comment by kmdalton** [comment]
oh dear. apparently there is no way to specify uncertainy keys manually, and careless expects that the key to have the same name as the intensity key plus a prefix like "SIG" or "Sig". if you change the column name from "SIGMA" to "SIGIOBS", it should work fine. I will file a separate issue for this. it is a bug.

**Comment by kmdalton** [comment]
@DorisMai, your issue will be fixed by #104 
Thanks for the feeback!
